### PR TITLE
Feature: Assigning IP and Gateway before the container running

### PR DESCRIPTION
**Pull request by ghost**
Config container's ip and gateway in this way :  docker run --ip-address=10.10.10.10 --gateway=10.10.254.254 ubuntu ping github.com

### CODE REVIEW

**Comment by jfrazelle** [comment]
Can you please sign your commits following these rules:

https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work

The easiest way to do this is to amend the last commit:

`git commit --amend -s --no-edit && git push -f`

**Comment by iddmx** [comment]
Edit New issue
Feature: Assigning IP and Gateway before the container runninghttps://github.com/docker/docker/issues/9972

**Comment by iddmx** [comment]
Feature: Assigning IP and Gateway before the container running
https://github.com/docker/docker/issues/9972

This is my first time to pull requests. I'm sorry that I'm not familiar with the Contribution Guidelines. I am trying to fix my fault.  I have amend my commit, but this page didn't change, I have no idea what should I do.

**Comment by MalteJ** [comment]
not bad!

Can you add some documentation?
docs/man/docker-run.md.1
docs/sources/articles/networking.md
docs/sources/reference/api/docker_remote_api_v1.17.md
docs/sources/reference/commandline/cli.md
docs/sources/reference/run.md

Slowly we get a lot of network related configuration flags. Too many?

**Comment by iddmx** [comment]
It's ok. But I have to sleep now, I can do that tomorrow. Will my code merge to the master ?

**Comment by MalteJ** [comment]
maybe it's a good idea to set a CIDR instead of an IP? e.g. 10.42.23.1/16

**Comment by iddmx** [comment]
You're right ! I will change that tomorrow or the day after tomorrow.

File: daemon/networkdriver/bridge/driver.go [hunk 1]
```diff
@@ -379,6 +381,13 @@ func Allocate(job *engine.Job) engine.Status {
 		return job.Error(err)
 	}
 
+	// If no explicit gateway was given, use default.
+	if requestedGateway != nil {
```
Comment (LK4D4) [inline daemon/networkdriver/bridge/driver.go]: I personally don't like unnecessary vars and elses :)
so I prefer such code looks like:
```
gateway := bridgeNetwork.IP
if requestedGateway != nil {
    gateway = requestedGateway
}
```

**Comment by LK4D4** [comment]
@iddmx thanks for contribution

**Comment by LK4D4** [comment]
Also this deserves `integration-cli` test

**Comment by eolamey** [comment]
For reference, there already is an opened pull request for setting a container's default gateway: https://github.com/docker/docker/pull/9381.

**Comment by MalteJ** [comment]
#9381 sets the gateway globally for all containers. It is a Daemon parameter.
This PR sets the gateway for one particular container. It is a run parameter.

**Comment by phemmer** [comment]
I submitted a much more flexible implementation for this feature (#7436) and it was rejected...

**Comment by LK4D4** [comment]
@phemmer Don't worry, maybe this will be rejected too :)

File: daemon/networkdriver/bridge/driver.go (hunk 1 above)
Comment (iddmx) [inline daemon/networkdriver/bridge/driver.go]: I will change my code and add some documentation. Anyway, thanks for code review :)

**Comment by resouer** [comment]
According #7436, @phemmer 's job is much more reasonable, but @shykes has defined the principle of this stuff:
"The Docker API should be used to consume infrastructure - not define it"

Google it & propose it in mailing list before sending a PR will save u a lot of time.

**Comment by crosbymichael** [comment]
Closing but please take a look at the network proposal that is currently being worked on by other maintainers and people in the community.  It will allow you to customize the network more and it would be good to have your input on the proposal to make sure your use case is known. 

#9983

**Pull request closed without merging**